torch/distributions/categorical.py

```diff
@@ -67,7 +67,12 @@ def sample(self, sample_shape=torch.Size()):
 
     def log_prob(self, value):
         self._validate_log_prob_arg(value)
-        param_shape = value.size() + self.probs.size()[-1:]
+        if self.batch_shape:
```

Comment:
I believe `self.batch_shape == self.probs.size()[:-1]` but is cheaper to compute if `logits` was specified. Consider replacing this with
```py
batch_shape = self.batch_shape
value_shape = torch._C.infer_size(value.shape, batch_shape) if batch_shape else value.shape
param_shape = value_shape + self._dim
```
and defining `self._dim` in the `__init__` method:
```py
self._dim = self._logits.shape[-1:] if self._probs is None else self._probs[-1:]
```

Reply:
Or since you definitely need logits below, simply define logits as a local variable and query it for shape.